### ISSUE TITLE
Camera fails to show sprites outside visible size

### ISSUE THREAD

**Issue by Tiger-Coding** [cocos2d/cocos2d-x#10487]
From the forum:
http://discuss.cocos2d-x.org/t/camera-bug-when-moving/19904/2

Using 3.4 final, I create a bunch of sprites, then position them one by one to the right.  I move the camera to the right, and everything off the screen is hidden.  However, if I move the layer to the left, I can see the sprites.

Code from forum post:

The init of HelloWorld:

```
auto spritebatch = SpriteBatchNode::create("MySprites.png");
addChild(spritebatch);
auto spriteCache = SpriteFrameCache::getInstance();
spriteCache->addSpriteFramesWithFile("MySprites.plist");
```

Then I create the sprites and add them to the layer, each one positioned to the right of the previous sprite.

```
for (int i = 0; i < 100; i++) {
    auto sprite = Sprite::createWithSpriteFrameName("MySprite.png");
    sprite->setPosition(Vec2(visibleSize.width/2 + origin.x + (visibleSize.width * i), visibleSize.height/2 + origin.y));
    this->addChild(sprite);
}
```

In the onEnter:

```
void HelloWorld::onEnter()
{
Layer::onEnter();
Size visibleSize = Director::getInstance()->getVisibleSize();
Camera *camera = Director::getInstance()->getRunningScene()->getCameras().front();
auto move = MoveBy::create(35.0, Vec3(10 * visibleSize.width, 0, 0));
camera->runAction(move);
}
```

Related?
https://github.com/cocos2d/cocos2d-x/issues/10275

**Comment by marianopixowl** [comment]
I've experienced this issue too in cocos2d-x 3.4. It seems as if culling is not taking into account the camera movement, but only the original screen bounds. (Looking at _Renderer::checkVisibility()_ seems to confirm this. I can't see the camera being used in the visibility calculations).

For the time being, I'm setting _#define CC_USE_CULLING 0_ as a workaround. (Which by the way, creates a compilation error in Label::draw() since the code references a local variable only defined within the _#if CC_USE_CULLING ... #endif_ block).

**Comment by zeyangl** [comment]
confirm issue exists in 3.6...

**Comment by fusijie** [comment]
I have added a test case to reproduce this bug.

Renderer->Auto-culling is affected by transform and view matrix. Now the implementation ignore the view matrix dirty so lead to this bug.

The `checkVisibility` method is fixed [here](https://github.com/cocos2d/cocos2d-x/commit/4b768c5fd81a5cfd6b0e4e1fc5210630a7f8e75e) by @yangws ?

But `draw` method still not check view matrix dirty. Let's discuss this issue here. @yangws @dabingnn @super626

**Comment by Cybertwip** [comment]
This happens also when drawing a FastTMXTiledMap and its TMXLayer objects.

This seems to partially solve the camera cull for TMX Layers:

```
void TMXLayer::draw(Renderer *renderer, const Mat4& transform, uint32_t flags)
{
    _quadsDirty = true;

    updateTotalQuads();

//    if( flags != 0 || _dirty || _quadsDirty )
//    {
//        Size s = Director::getInstance()->getWinSize();
//        auto rect = Rect(0, 0, s.width, s.height);

        auto director = Director::getInstance();

        Rect visibleRect(Camera::getVisitingCamera()->getPositionX() - director->getVisibleSize().width * 0.5, 
                         Camera::getVisitingCamera()->getPositionY() - director->getVisibleSize().height * 0.5,
                         director->getVisibleSize().width, 
                         director->getVisibleSize().height);

        Mat4 inv = transform;
        inv.inverse();
        visibleRect = RectApplyTransform(visibleRect, inv);

        updateTiles(visibleRect);
        updateIndexBuffer();
        updatePrimitives();
        _dirty = false;
//    }

    if(_renderCommands.size() < static_cast<size_t>(_primitives.size()))
    {
        _renderCommands.resize(_primitives.size());
    }

    int index = 0;
    for(const auto& iter : _primitives)
    {
        if(iter.second->getCount() > 0)
        {
            auto& cmd = _renderCommands[index++];
            cmd.init(iter.first, _texture->getName(), getGLProgramState(), BlendFunc::ALPHA_NON_PREMULTIPLIED, iter.second, _modelViewTransform, flags);
            renderer->addCommand(&cmd);
        }
    }
}
```

**Comment by fusijie** [comment]
**Here is auto-culling's performance test.**
<img width="1019" alt="peformancetest1" src="https://cloud.githubusercontent.com/assets/5085300/8975376/19a4f682-36b0-11e5-89a8-19510c9b5b6a.png">

**Result:**
- All inside of screen
  - Enable or disable auto-culling have almost the same performance.
  - But if `not check dirty`, will cause perfermance loss in Android (-20%).
- All outside of screen
  - Enable auto-culling is much better than disable it.(iOS 105%, Android 15%)
- Half inside of screen and half outside of screen
  - Enable auto-culling is much better than disable it.(iOS 26%, Android 18%)

**Conclusion:**
- Keep enable auto-culling.
- As far as possible, enable check transform dirty.
- For fix this issue:
  - Enable check dirty for `default camera` to ensure 2d performance.
  - Disable check dirty for other cameras.

**Remaining issue:**
- Performance test.
- To keep trace with all cameras' viewMatrix changing.

**Comment by fusijie** [comment]
According to the performance test, fixed in #13105.
New2 Code -> #13105.
<img width="1105" alt="peformancetest2" src="https://cloud.githubusercontent.com/assets/5085300/8975557/0aa55b4c-36b3-11e5-8fec-532650c70f47.png">
